### DISCUSSION TITLE
Adding a hook in ObjectModel constructor to allow a module to register additionnal properties

### DISCUSSION THREAD

**Discussion by mytrain** [PrestaShop/PrestaShop#35147] [Ideas]
### The problem

Adding extra field in a table (product, address, ...) is quite painfull and extra field are not taken into account by other modules when copying or duplicating object. This is particulary true in the case of addresses.

### The solution

Adding the following Hook in the ObjectModel::___construct 
<pre>
 public function __construct($id = null, $id_lang = null, $id_shop = null, $translator = null)
    {
        $class_name = get_class($this);

        //New Hook
        Hook::exec('actionObject' . $this->getFullyQualifiedName() . 'Construct', ['object' => $this]);

        if (!isset(ObjectModel::$loaded_classes[$class_name])) {
            $this->def = ObjectModel::getDefinition($class_name);
. . .
</pre>

and the following method
<pre>
        public function addModuleSpecificField($name,$definition,$value) {
                if (!isset($this->{$name})) {
                        $classname = get_class($this);
                        $classname::$definition['fields'][$name] = $definition;
                        $this->{$name} = $value;
                }
        }

</pre>

will allow a module to register extra field (exemple of adding field dmeori_eori_number in Address model
<pre>
    public function hookActionObjectAddressConstruct($params)
    {
		$object = $params['object'];
		$object->addModuleSpecificField("dmeori_eori_number",array('type' => ObjectModel::TYPE_STRING),null);
    }

	
    public function hookActionCustomerAddressFormBuilderModifier($params) 
    {
		
        $formBuilder = $params['form_builder'];

        $formBuilder->add('dmeori_eori_number', Symfony\Component\Form\Extension\Core\Type\TextType::class, [
            'label' => $this->l('Numéro EORI'),
            'required' => false,
        ]);
		
        $address = new Address($params['id']);
		$params['data']['dmeori_eori_number'] = $address->dmeori_eori_number;
        $formBuilder->setData($params['data']);
    }

    public function hookActionAfterUpdateCustomerAddressFormHandler(array $params)
    {
		DMTools::logDM(json_encode($params));
		$address = new Address((int)$params['id']);
		$address->dmeori_eori_number = $params['form_data']['dmeori_eori_number'];
		$address->update();
    }

    public function hookActionAfterCreateCustomerAddressFormHandler(array $params)
    {
		DMTools::logDM(json_encode($params));
		$address = new Address((int)$params['id']);
		$address->dmeori_eori_number = $params['form_data']['dmeori_eori_number'];
		$address->update();
    } 
	
    public function hookAdditionalCustomerAddressFields($params)
    {
		$address = new Address(Tools::getValue('id_address'));
        return [
            (new FormField)
                ->setName('dmeori_eori_number')
                ->setType('text')
                ->setRequired(false) // is Required
                ->setValue($address->dmeori_eori_number )
				->setLabel($this->l('Numéro EORI'))
		];
    }
</pre>

### Additional context

I'm not familliar with pulling request in the code so if somebody else is interested fell free to do it.

### Do you plan to work on this subject?

- [ ] I'm willing to contribute a [formal specification](https://github.com/PrestaShop/prestashop-specs).
- [ ] I'm willing to provide any wireframes or design assets required for this feature.
- [ ] I'm willing to submit a Pull Request that implements this feature.
- [ ] I'm willing to help verify that the implemented feature works as intended and produces no unintended side effects.

**Comment by matks** [comment]
Hello @mytrain I'm afraid we don't want to do this :/

I explain: it looks like a good idea but altering the ObjectModel structure is a dangerous method. Because it makes it a lot harder to perform shop upgrade.

For example maybe between version 8 and version 9 we want to alter the columns of a given ObjectModel. However using the hook you might have already altered its structure, making the upgrade process fail.

It is rather suggested to create new SQL tables and connect them to existing ObjectModel structures using foreign keys.

For example for your usecase dmeori_eori_number instead of adding it inside the Address Model I would suggest to create a new DMeori_Address model and create a relationship between address and your new model. This makes sure the Address model remains unaltered and eases upgrading the SQL structure, preventing you from going into customization nightmare where you have altered so much PrestaShop you cannot upgrade to newer versions.

**Comment by mytrain** [comment]
Mathieu,

Thanks for your return and your argument worth it as presatshop native code shall be upgradable whatever module implementation.

My main purpose is to be able to add extra fields in a model and be sure they will be corretly handled by any prestashop (or thrid party module) correctly. 
As exemple if we consider the Address model delivered by Prestashop a class such as **src/Adapter/Address/CommandHandle/EditCustomerAddressHandler**. The code of the class shall not be modified in case of extra field added by a thrid party module (especially when it copies the address in function **handle**). 
Using an extra table and foreign key will not allow the  EditCustomerAddressHandler::handle code to copy automatically information in thrid party module tables and the newly created instance will therefore be incomplete from thrid party module point of view.

In my proposal we can prefix new fields with module name or (nsp_ for non prestashop) in order to avoid any overlapping with prestashop future field.

But maybe another design may satisfy my design ?

Bien à vous.

**Comment by kpodemski** [comment]
> My main purpose is to be able to add extra fields in a model and be sure they will be corretly handled by any prestashop (or thrid party module) correctly.

Since we are not using foreign keys, I'm afraid you need to make the lives of other developers easier by providing them with information on how to support your modules instead of aiming to have automatic support from scratch.

I think all project members will agree with me that:
- you should never edit the core tables
- you should never alter properties of native models
- you should build on top and extend instead of modifying directly :)

I am pretty sure that similar issues are in other platforms as well. Extensibility comes with some downsides.

**Comment by mcaldex** [comment]
Hi everyone,
as a module developer, I often find myself extending PrestaShop's core models.
For example I often need to add additional custom properties attached to Addresses, which I can show both in frontoffice and backoffice.

I will explain the best way I've found in order to do that without alterting PrestaShop's core database structure (I don't think my method is perfect but I think that we could replace some functions overrides by introducing a couple of native prestashop hooks):

suppose we want to extend the Address object: we create a module with a custom db table which we will use to host the additional Address data. We put the primary key as an unsigned integer "id_address" (same as the PrestaShop's one which we will then link)

so we create an override extending AddressCore class:

```
<?php

class Address extends AddressCore
{
  public const additionalfields_table = 'hoscustomregistration_additionaladdressfields';
  public const additionalfields = ['contact_email'];
  private $additionalfieldsLoaded = false;
  public $contact_email;
  
  public function __construct($id_address = null, $id_lang = null)
  {
      parent::__construct($id_address, $id_lang);
      $this->loadAdditionalfieldsFromDb();
  }
  
  public function loadAdditionalfieldsFromDb($force_reload = false)
  {
      if($force_reload){
          $this->additionalfieldsLoaded = false;
      }
      if($this->additionalfieldsLoaded)
          return;
      if($this->id){
          $this->additionalfieldsLoaded = true;
          if(
              $custom_fields = Db::getInstance()->getRow("SELECT ".implode(',',self::additionalfields)." FROM "._DB_PREFIX_.self::additionalfields_table." WHERE id_address = $this->id")
          ){
              foreach($custom_fields as $key => $value){
                  $this->$key = $value;
              }
          }
      }
  }
  
  public function validateFields($die = true, $error_return = false)
  {
      //we execute validation for core fields
      if(($result = parent::validateFields($die,$error_return)) !== true){
          return $result;
      }
      //we can then execute our additional validations for our own additional fields
  
      return true;
  }
}
```

as you can see, we're declaring our new public property contact_email, which we will load upon object creation using the constructor : if we can add a native hook on the ObjectModel class like @mytrain is saying, we could move this logic into our module, getting rid of it inside the override.
we're also adding additional validation for our own properties by overriding the validateFields method -> again, we could get rid of this override only if we implement another native hook on ObjectModel

in order to save our additional data in our custom table, we can register hooks in our module for doing this:

```
public function hookActionObjectAddressAddAfter($params){
    $object = $params['object'];
    return $this->updateCustomerAddressCustomFields($object);
}

public function hookActionObjectAddressUpdateAfter($params){
    $object = $params['object'];
    return $this->updateCustomerAddressCustomFields($object);
}

private function updateCustomerAddressCustomFields($object){
    $custom_field_names = Address::additionalfields;
    $custom_fields = [];
    foreach($custom_field_names as $custom_field_name){
        $custom_fields[$custom_field_name] = pSQL($object->$custom_field_name);
    }
    $custom_fields['id_address'] = $object->id;
    return Db::getInstance()->insert(Address::additionalfields_table,$custom_fields,true,true,Db::ON_DUPLICATE_KEY);
}
```

and there we go.
@kpodemski could we consider to add a couple of native hooks in prestashop core to do this? On the longer run, we should try to minimize overrides, and to do this we need to implement more hooks.
Please also consider to document this pattern in the dev docs after we implement the hooks

have a nice day, I hope I have helped somehow
